### PR TITLE
perf(http): optimize file server

### PR DESCRIPTION
**Pull request by littledivy**
Towards #2107 

Major changes:

- use native readable streams optimization. `Deno.File#readable` instead of a wrapper stream. See https://github.com/denoland/deno/pull/14284
- use fnv-1a default etag hash function. Adds an option to allow using others.
- ~~faster hex function for out of other algorithms (like `sha-1`), using hex lookup table.~~
- disables logging by default when invoked independently. Replaces the `--quiet` flag with `--verbose`. Public API remains unchanged.
- ~~use simpler extname extraction to avoid scanning manually. (downsides?)~~

Benchmarks:

`main`:
```
Running 10s test @ http://localhost:4507/README.md
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.44ms  522.77us  26.73ms   97.55%
    Req/Sec     1.40k   119.07     2.63k    97.39%
  112288 requests in 10.10s, 432.42MB read
Requests/sec:  11114.28
Transfer/sec:     42.80MB
```

This patch:
```
wrk -t8 -c20 -d10s http://localhost:4507/README.md
Running 10s test @ http://localhost:4507/README.md
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.01ms  163.42us   6.32ms   85.73%
    Req/Sec     1.99k    67.51     2.09k    86.51%
  160068 requests in 10.10s, 611.53MB read
Requests/sec:  15848.72
Transfer/sec:     60.55MB
```

### CODE REVIEW

File: http/file_server.ts [hunk 1]
```diff
@@ -42,7 +42,8 @@ interface FileServerArgs {
   quiet: boolean;
 }
 
-const encoder = new TextEncoder();
+// @ts-ignore `Deno.core` is not a public API
+const encode = Deno.core.encode || new TextEncoder().encode;
```
Comment (lucacasonato) [inline http/file_server.ts:46]: Not in favor of this one.

File: http/file_server.ts [hunk 2]
```diff
@@ -174,20 +175,51 @@ const MEDIA_TYPES: Record<string, string> = {
 
 /** Returns the content-type based on the extension of a path. */
 function contentType(path: string): string | undefined {
-  return MEDIA_TYPES[extname(path)];
+  const ext = path.split(".").pop();
```
Comment (lucacasonato) [inline http/file_server.ts:178]: Optimize extname instead.

File: http/file_server.ts [hunk 3]
```diff
@@ -174,20 +175,51 @@ const MEDIA_TYPES: Record<string, string> = {
...
+  for (let i = 0; i < buf.length; i++) {
+    hash ^= buf.charCodeAt(i);
+    // Equivalent to `hash *= 16777619` without using BigInt
+    // 32-bit FNV prime
+    hash += (hash << 1) + (hash << 4) + (hash << 7) + (hash << 8) +
+      (hash << 24);
+  }
+  // 32-bit hex string
+  return (hash >>> 0).toString(16);
+}
```
Comment (lucacasonato) [inline http/file_server.ts:208]: Is this not available is std/crypto?

File: http/file_server.ts [hunk 4]
```diff
@@ -174,20 +175,51 @@ const MEDIA_TYPES: Record<string, string> = {
...
+  hexLookup[n] = hexOctet;
+}
+
+function hex(buf: ArrayBuffer): string {
+  const u8arr = new Uint8Array(buf);
+  const len = u8arr.length;
+  let hexOctets = "";
+  for (let i = 0; i < len; ++i) hexOctets += hexLookup[u8arr[i]];
+  return hexOctets;
+}
```
Comment (lucacasonato) [inline http/file_server.ts:194]: Use std/encoding/hex and put this optimized encoder in there

File: testing/asserts.ts [hunk 5]
```diff
@@ -97,7 +97,9 @@ function buildMessage(
         ? createColor(detail.type, { background: true })(detail.value)
         : detail.value
     ).join("") ?? result.value;
-    diffMessages.push(c(`${createSign(result.type)}${line}`));
+    diffMessages.push(c(`${
+      createSign(result.type)
+    }${line}`));
```
Comment (lucacasonato) [inline testing/asserts.ts:102]: Unrelated change?

**Review by lucacasonato — COMMENTED** [review]
Great that you have managed to optimize this. This is code we want people to read and understand though, so we shouldn't duplicate functionality from other std/ modules.

File: testing/asserts.ts (hunk 5 above)
Comment (bartlomieju) [reply testing/asserts.ts:102]: I've seen this change recently in some other PR, seems to be related to dprint. CC @dsherret

File: http/file_server.ts (hunk 1 above)
Comment (bartlomieju) [reply http/file_server.ts:46]: Agreed, we should just use `TextEncoder`

File: http/file_server.ts (hunk 3 above)
Comment (littledivy) [reply http/file_server.ts:208]: nope, can't find it

File: http/file_server.ts (hunk 4 above)
Comment (littledivy) [reply http/file_server.ts:194]: Didn't know about encoding/hex, it is equivalent in terms of performance. Though we should add a Uint8Array -> hex string function.

File: http/file_server.ts (hunk 1 above)
Comment (littledivy) [reply http/file_server.ts]: fine, comes with a performance penalty but whatever. Updated the benchmark results.

File: http/file_server.ts (hunk 2 above)
Comment (littledivy) [reply http/file_server.ts]: extname is fine actually, reverted.

**Review by bartlomieju — APPROVED** [review]
LGTM

File: http/file_server.ts [hunk 6]
```diff
@@ -233,16 +251,20 @@ function fileLenToString(len: number): string {
...
 ): Promise<Response> {
-  const [file, fileInfo] = await Promise.all([
-    Deno.open(filePath),
-    Deno.stat(filePath),
-  ]);
-
+  const file = await Deno.open(filePath);
+  if (fileInfo === undefined) {
+    fileInfo = await Deno.stat(filePath);
+  }
```
Comment (lucacasonato) [inline http/file_server.ts:267]: This serializes the calls. You can still have them be parallel while adding this optimization.

File: http/file_server.ts [hunk 7]
```diff
@@ -233,16 +251,20 @@ function fileLenToString(len: number): string {
...
  * @param filePath Path of the file to serve.
+ * @param etagAlgorithm The algorithm to use for generating the ETag. Defaults to "fnv1a".
+ * @param fileInfo An optional FileInfo object returned by Deno.stat. It is used
+ * for optimization purposes.
  */
 export async function serveFile(
   req: Request,
   filePath: string,
+  etagAlgorithm: EtagAlgorithm = "fnv1a",
+  fileInfo?: Deno.FileInfo,
```
Comment (lucacasonato) [inline http/file_server.ts:262]: These two should be an option bag according to our style guide I think (public api)

**Review by lucacasonato — APPROVED** [review]
LGTM!
